### PR TITLE
Bug/ipos 2476 error handling

### PR DESCRIPTION
**Pull request by jake-fooda**
WIP

### CODE REVIEW

File: graphql-ios/Service/GraphQLAuthentication.swift [hunk 1]
```diff
@@ -7,5 +7,6 @@
 //
 
 public enum GraphQLAuthentication {
-    case authenticated, anonymous
+    case authenticated(sessionToken: String)
+    case anonymous(sessionToken: String?)
```
Comment (blakemacnair) [inline graphql-ios/Service/GraphQLAuthentication.swift]: Shouldn't we force prevent a user from sending any session token with any `anonymous` case call?

File: graphql-ios/Service/GraphQLClient.swift [hunk 2]
```diff
@@ -174,45 +162,38 @@ private extension GraphQLClient {
                 headers[key] = value
             }
         }
-
+
```
Comment (blakemacnair) [inline graphql-ios/Service/GraphQLClient.swift]: 👀 👀 👀 👀 👀 👀 👀 👀

File: graphql-ios/Service/GraphQLAuthentication.swift (hunk 1 above)
Comment (jake-fooda) [reply graphql-ios/Service/GraphQLAuthentication.swift]: Existing logic for GraphQL (and REST) had us sending the `sessionToken` if we had it.  I *think* the thought here is that if we have this data it doesn't hurt to send.

File: graphql-ios/Service/GraphQLClient.swift (hunk 2 above)
Comment (jake-fooda) [reply graphql-ios/Service/GraphQLClient.swift]: lol no linter!

File: graphql-ios/Service/GraphQLClient.swift [hunk 3]
```diff
@@ -158,12 +144,14 @@ private extension GraphQLClient {
                                                  "total": response.timeline.totalDuration],
                                     "response": rawJson ?? [:],
                                     "error": (error as? GraphQLRemoteError)?.debugDescription ?? error.localizedDescription
-                ])
+            ])
             completion(.failure(error))
         }
     }
 
-    func requestHeaders(with customHeader: [String: String]?, clientToken: String?, authentication: GraphQLAuthentication) throws -> [String: String] {
```
Comment (blakemacnair) [inline graphql-ios/Service/GraphQLClient.swift]: I don't think this func needs to `throw` anymore does it?

Comment (jake-fooda) [reply graphql-ios/Service/GraphQLClient.swift]: nice catch 👍

**Comment by jake-fooda** [comment]
@blakemacnair @MrNiceMrGood this is ready for review with tests now.  @MrNiceMrGood didn't realize you weren't added to this repo... added now.

File: graphql-iosTests/Service/GraphQLClientTests.swift [hunk 4]
```diff
@@ -9,13 +9,27 @@
 import XCTest
 @testable import graphql_ios
 
+struct MockResponse: Decodable {
+
+}
+
+struct UnexpectedJsonResponse: Decodable {
```
Comment (blakemacnair) [inline graphql-iosTests/Service/GraphQLClientTests.swift]: Could you add a small comment to help show what this little struct is for?

File: graphql-iosTests/Service/GraphQLClientTests.swift [hunk 5]
```diff
@@ -9,13 +9,27 @@
 import XCTest
 @testable import graphql_ios
 
+struct MockResponse: Decodable {
```
[the remark this answers is not in the document]
Comment (jake-fooda) [reply graphql-iosTests/Service/GraphQLClientTests.swift]: Added comments 👍

**Comment by jake-fooda** [comment]
@blakemacnair pushed https://github.com/Fooda/graphql-ios/pull/6/commits/2ce0d7dbe68179f4a4530eed1359933bb54f4867 after the approval.  Mind taking a look at that commit?

**Pull request merged**